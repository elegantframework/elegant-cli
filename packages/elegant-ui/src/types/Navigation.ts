/**
 * A section of url links.
 */
export interface NavigationSection {
    /**
     * The title of this navigation section.
     */
    title: string;
    /**
     * A list of url navigation links.
     */
    links: NavigationLink[];
};

/**
 * A url link with a title.
 */
export interface NavigationLink {
    /**
     * The title of the URL.
     */
    title: string;
    /**
     * The actual url link.
     */
    href: string;
    /**
     * Is this url to an external website?
     */
    external?: boolean;
};