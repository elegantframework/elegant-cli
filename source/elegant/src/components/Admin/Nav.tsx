"use client";
import Link from "next/link";
import {
  ArrowLeft,
  BarChart3,
  Edit3,
  LayoutDashboard,
  Menu,
  Newspaper,
  Settings,
  FileCode,
  Github,
  WalletCards,
} from "lucide-react";
import {
  useParams,
  usePathname,
  useSelectedLayoutSegments,
} from "next/navigation";
import { ReactNode, useEffect, useMemo, useState } from "react";
import { ElegantLogo } from "@brandonowens/elegant-ui";

const externalLinks = [
  {
    name: "Star on GitHub",
    href: "https://github.com/elegantframework/elegant-cli",
    icon: <Github width={18} />,
  },
  {
    name: "Documentation",
    href: "https://www.elegantframework.com/docs/installation",
    icon: <FileCode width={18} />,
  },
];

export default function Nav({ children }: { children: ReactNode }) {
  const segments = useSelectedLayoutSegments();
  const pathname = usePathname();
  const { id } = useParams() as { id?: string };

  const [siteId, setSiteId] = useState<string | null>();

  useEffect(() => {
    // if (segments[0] === "post" && id) {
    //   getSiteFromPostId(id).then((id) => {
    //     setSiteId(id);
    //   });
    // }
  }, [segments, id]);

  const tabs = useMemo(() => {
    if (segments[0] === "site" && id) {
      return [
        {
          name: "Back to All Sites",
          href: "/sites",
          icon: <ArrowLeft width={18} />,
        },
        {
          name: "Posts",
          href: `/site/${id}`,
          isActive: segments.length === 2,
          icon: <Newspaper width={18} />,
        },
        {
          name: "Collections",
          href: `/site/${id}/analytics`,
          isActive: segments.includes("analytics"),
          icon: <BarChart3 width={18} />,
        },
        {
          name: "Settings",
          href: `/admin/settings`,
          isActive: segments.includes("/admin/settings"),
          icon: <Settings width={18} />,
        },
      ];
    } else if (segments[0] === "post" && id) {
      return [
        {
          name: "Back to All Posts",
          href: siteId ? `/site/${siteId}` : "/sites",
          icon: <ArrowLeft width={18} />,
        },
        {
          name: "Editor",
          href: `/post/${id}`,
          isActive: segments.length === 2,
          icon: <Edit3 width={18} />,
        },
        {
          name: "Settings",
          href: `/admin/settings`,
          isActive: segments.includes("/admin/settings"),
          icon: <Settings width={18} />,
        },
      ];
    }
    return [
      {
        name: "Overview",
        href: "/admin",
        isActive: pathname === "/admin",
        icon: <LayoutDashboard width={18} />,
      },
      {
        name: "Posts",
        href: `/site/${id}`,
        isActive: segments.length === 2,
        icon: <Newspaper width={18} />,
      },
      {
        name: "Collections",
        href: `/admin/collections`,
        isActive: pathname.includes("collections"),
        icon: <WalletCards width={18} />,
      },
      {
        name: "Settings",
        href: "/admin/settings",
        isActive: pathname === "/admin/settings",
        icon: <Settings width={18} />,
      },
    ];
  }, [segments, id, siteId, pathname]);

  const [showSidebar, setShowSidebar] = useState(false);

  useEffect(() => {
    // hide sidebar on path change
    setShowSidebar(false);
  }, [pathname]);

  return (
    <>
      <button
        className={`fixed z-20 ${
          // left align for Editor, right align for other pages
          segments[0] === "post" && segments.length === 2 && !showSidebar
            ? "left-5 top-5"
            : "right-5 top-7"
        } sm:hidden`}
        onClick={() => setShowSidebar(!showSidebar)}
      >
        <Menu width={20} />
      </button>
      <div
        className={`transform ${
          showSidebar ? "w-full translate-x-0" : "-translate-x-full"
        } fixed z-10 flex h-full flex-col justify-between border-r border-stone-200 bg-stone-100 p-4 transition-all dark:border-stone-700 dark:bg-stone-900 sm:w-60 sm:translate-x-0`}
      >
        <div className="grid gap-2">
          <div className="flex items-center space-x-2 rounded-lg px-2 py-1.5">
            <Link
                href="/admin"
                className="mr-3 flex-none w-[2.0625rem] overflow-hidden"
            >
                <ElegantLogo className="w-auto h-7"/>
            </Link>
          </div>
          <div className="grid gap-1 mt-1">
            {tabs.map(({ name, href, isActive, icon }) => (
              <Link
                key={name}
                href={href}
                className={`flex items-center space-x-3 ${
                  isActive ? "bg-stone-200 text-black dark:bg-stone-700" : ""
                } rounded-lg px-2 py-1.5 transition-all duration-150 ease-in-out hover:bg-stone-200 active:bg-stone-300 dark:text-white dark:hover:bg-stone-700 dark:active:bg-stone-800`}
              >
                {icon}
                <span className="text-sm font-medium">{name}</span>
              </Link>
            ))}
          </div>
        </div>
        <div>
          <div className="grid gap-1">
            {externalLinks.map(({ name, href, icon }) => (
              <a
                key={name}
                href={href}
                target="_blank"
                rel="noopener noreferrer"
                className="flex items-center justify-between rounded-lg px-2 py-1.5 transition-all duration-150 ease-in-out hover:bg-stone-200 active:bg-stone-300 dark:text-white dark:hover:bg-stone-700 dark:active:bg-stone-800"
              >
                <div className="flex items-center space-x-3">
                  {icon}
                  <span className="text-sm font-medium">{name}</span>
                </div>
                <p>â†—</p>
              </a>
            ))}
          </div>
          <div className="my-2 border-t border-stone-200 dark:border-stone-700" />
          {children}
        </div>
      </div>
    </>
  );
}
