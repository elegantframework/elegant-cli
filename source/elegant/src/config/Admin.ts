/**
 * The Elegant admin panel configuration.
 * Our config is bootstrapped from our .env values.
 * If no .env file is found, it will use the default values set here.
 */
const AdminConfig = () => {
    return {
        /*
        |--------------------------------------------------------------------------
        | CMS Panel Name
        |--------------------------------------------------------------------------
        |
        | This value is the name displayed in the title of your CMS panel.
        |
        */
        'cms_name': process.env.NEXT_PUBLIC_CMS_NAME || 'Elegant',

        /*
        |--------------------------------------------------------------------------
        | CMS Repository Branch
        |--------------------------------------------------------------------------
        |
        | This is the repository branch where your cms content will be created and stored.
        |
        |
        */
        'cms_repository_branch': process.env.NEXT_PUBLIC_CMS_REPOSITORY_BRANCH || 'main',

        /*
        |--------------------------------------------------------------------------
        | CMS GitHub ID
        |--------------------------------------------------------------------------
        |
        | This is the GitHub Client ID of your GitHub Oauth application.
        |
        |
        */
        'cms_github_id': process.env.NEXT_PUBLIC_CMS_GITHUB_ID || '',

        /*
        |--------------------------------------------------------------------------
        | CMS GitHub Secret
        |--------------------------------------------------------------------------
        |
        | This is the GitHub Client secret of your  
        |
        |
        */
        'cms_github_secret': process.env.NEXT_PUBLIC_CMS_GITHUB_SECRET || '',

        /*
        |--------------------------------------------------------------------------
        | CMS GitHub Token Secret
        |--------------------------------------------------------------------------
        |
        | This is a random string with at least 32 characters. 
        | You can generate a random string using the service below.
        |
        | https://onlinehashtools.com/generate-random-sha256-hash?&count=1
        |
        |
        */
        'cms_token_secret': process.env.NEXT_PUBLIC_CMS_TOKEN_SECRET || '',

        /*
        |--------------------------------------------------------------------------
        | CMS Repository Slug
        |--------------------------------------------------------------------------
        |
        | This is the slug of the repository where you store your content.
        | If empty AND on Vercel, this will default to VERCEL_GIT_REPO_SLUG.
        |
        |
        */
        'cms_repository_slug': process.env.NEXT_PUBLIC_CMS_REPOSITORY_SLUG || '',

        /*
        |--------------------------------------------------------------------------
        | CMS Repository Owner
        |--------------------------------------------------------------------------
        |
        | This is the owner of the repository where you store your content.
        | If the repository owner is not set, the GitHub username of the logged in account will be used.
        |
        */
        'cms_repository_owner': process.env.NEXT_PUBLIC_CMS_REPOSITORY_OWNER || '',

        /*
        |--------------------------------------------------------------------------
        | CMS Content Path
        |--------------------------------------------------------------------------
        |
        | This is the relative folder path to where you store your content in Markdown.
        |
        */
        'cms_content_path': process.env.NEXT_PUBLIC_CMS_CONTENT_PATH || '_content',

        /*
        |--------------------------------------------------------------------------
        | CMS Monorepo Path
        |--------------------------------------------------------------------------
        |
        | This is the relative folder to where your Elegant application is located,
        | only if your project is set up as a monorepo with multiple projects.
        |
        */
        'cms_monorepo_path': process.env.NEXT_PUBLIC_CMS_MONOREPO_PATH || '',

         /*
        |--------------------------------------------------------------------------
        | CMS Asset Path
        |--------------------------------------------------------------------------
        |
        | This is the relative folder path to where assets save from the CMS panel
        | are stored.
        |
        */
        'cms_asset_path': process.env.NEXT_PUBLIC_CMS_ASSET_PATH || 'public/images',
    };
};

export default AdminConfig;