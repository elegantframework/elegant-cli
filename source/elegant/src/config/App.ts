/**
 * The Elegant application configuration.
 * Our config is bootstrapped from our .env values.
 * If no .env file is found, it will use the default values set here.
 */
const AppConfig = () => {
    return {
        /*
        |--------------------------------------------------------------------------
        | Application Name
        |--------------------------------------------------------------------------
        |
        | This value is the name of your application. This value is used when the
        | framework needs to place the application's name in a notification or
        | any other location as required by the application or its packages.
        |
        */
        'name': process.env.NEXT_PUBLIC_APP_NAME || "Elegant",

        /*
        |--------------------------------------------------------------------------
        | Application Description
        |--------------------------------------------------------------------------
        |
        | This value is the description of your application. 
        |
        */
        'description': process.env.NEXT_PUBLIC_APP_DESCRIPTION || "A content creation framework for rapidly building beautiful and expressive web applications.",

        /*
        |--------------------------------------------------------------------------
        | Application Tagline
        |--------------------------------------------------------------------------
        |
        | This is your applications slogan. It is used throughout the app, in places such as the meta description.
        |
        */
        'tagline': process.env.NEXT_PUBLIC_APP_TAGLINE || "Elevate your creative content.",

        /*
        |--------------------------------------------------------------------------
        | Application URL
        |--------------------------------------------------------------------------
        |
        | This URL is used by Elegant to properly generate URLs during build.
        | The application url is also passed to the rich snippets schema data to set a canonical url.
        | You should set this to the root of where your application is running.
        |
        */
        'url': process.env.NEXT_PUBLIC_APP_URL || "https://elegantframework.com",

        /*
        |--------------------------------------------------------------------------
        | Application Repository
        |--------------------------------------------------------------------------
        |
        | The GitHub URL of the application source code.
        |
        */
        'repository': process.env.NEXT_PUBLIC_APP_REPOSITORY || "",

        /*
        |--------------------------------------------------------------------------
        | Application Type
        |--------------------------------------------------------------------------
        |
        | The type of Elegant application.
        |
        */
        'type': process.env.NEXT_PUBLIC_APP_TYPE || "Organization",

        /*
        |--------------------------------------------------------------------------
        | Twitter Handle
        |--------------------------------------------------------------------------
        |
        |
        |
        */
        'twitter_handle': process.env.NEXT_PUBLIC_APP_TWITTER_HANDLE || "",

        /*
        |--------------------------------------------------------------------------
        | Twitter Site Handle
        |--------------------------------------------------------------------------
        |
        |
        */
        'twitter_site_handle': process.env.NEXT_PUBLIC_APP_TWITTER_SITE_HANDLE || "",

        /*
        |--------------------------------------------------------------------------
        | Convert Kit Action URL
        |--------------------------------------------------------------------------
        |
        | Convert Kit action urls used to send newsletter signup's from your application.
        |
        */
        'convert_action_url': process.env.NEXT_PUBLIC_CONVERTKIT_ACTION_URL || "",

        /*
        |--------------------------------------------------------------------------
        | Google Analytics ID
        |--------------------------------------------------------------------------
        |
        | The unique tracking ID for a Google Analytics account.
        |
        */
        'google_analytics_id': process.env.NEXT_PUBLIC_GOOGLE_ANALYTICS_ID || "",

        /*
        |--------------------------------------------------------------------------
        | Facebook App ID
        |--------------------------------------------------------------------------
        |
        |
        |
        */
        'facebook_app_id': process.env.NEXT_PUBLIC_FACEBOOK_APP_ID || "",

        /*
        |--------------------------------------------------------------------------
        | Twitter URL
        |--------------------------------------------------------------------------
        |
        | The url for the Twitter page for this application. 
        | This url is used in the social schema data and rich snippets.
        | The rich snippets tell Google and other crawlers that this website is associated with this social media account.
        */
        'twitter_url': process.env.NEXT_PUBLIC_APP_TWITTER_URL || "",

        /*
        |--------------------------------------------------------------------------
        | Facebook URL
        |--------------------------------------------------------------------------
        |
        | The url for the Facebook page for this application. 
        | This url is used in the social schema data and rich snippets.
        | The rich snippets tell Google and other crawlers that this website is associated with this social media account.
        |
        */
        'facebook_url': process.env.NEXT_PUBLIC_APP_FACEBOOK_URL || "",

        /*
        |--------------------------------------------------------------------------
        | Instagram URL
        |--------------------------------------------------------------------------
        | The url for the Instagram page for this application. 
        | This url is used in the social schema data and rich snippets.
        | The rich snippets tell Google and other crawlers that this website is associated with this social media account.
        */
        'instagram_url': process.env.NEXT_PUBLIC_APP_INSTAGRAM_URL || "",

        /*
        |--------------------------------------------------------------------------
        | Youtube URL
        |--------------------------------------------------------------------------
        | The url for the YouTube page for this application. 
        | This url is used in the social schema data and rich snippets.
        | The rich snippets tell Google and other crawlers that this website is associated with this social media account.
        */
        'youtube_url': process.env.NEXT_PUBLIC_APP_YOUTUBE_URL || "",

        /*
        |--------------------------------------------------------------------------
        | LinkedIn URL
        |--------------------------------------------------------------------------
        | The url for the LinkedIn page for this application. 
        | This url is used in the social schema data and rich snippets.
        | The rich snippets tell Google and other crawlers that this website is associated with this social media account.
        */
        'linkedin_url': process.env.NEXT_PUBLIC_APP_LINKEDIN_URL || "",

        /*
        |--------------------------------------------------------------------------
        | Application Contact Email
        |--------------------------------------------------------------------------
        | The public contact email to be displayed to users. 
        | Currently only used on the Trademark Policy page.
        |
        */
        'contact_email': process.env.NEXT_PUBLIC_CONTACT_EMAIL || "info@elegantframework.com",
    };
};

export default AppConfig;