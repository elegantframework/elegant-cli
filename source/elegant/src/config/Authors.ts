import brandonOwensAvatar from '@/img/authors/brandonowens.jpg';

export const brandonOwens = {
  name: 'Brandon Owens',
  twitter: 'thebrandonowens',
  avatar: brandonOwensAvatar.src,
};