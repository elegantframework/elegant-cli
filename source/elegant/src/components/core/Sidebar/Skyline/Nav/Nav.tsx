import { useRouter } from 'next/router';
import { NavigationSection } from '@/types/Navigation';
import clsx from 'clsx';
import NavItem from '../NavItem/NavItem';

interface Props {
    /**
     * Is this view for a mobile device?
     */
    mobile?: boolean;
    /**
     * The navigation list.
     */
    nav: NavigationSection[];
};

/**
 * The navigation menu to be used within the sidebar.
 * @returns An html ul navigation list.
 */
export default function Nav({
    mobile = false,
    nav
}: Props) {
    const router = useRouter();
    
    return(
        <nav className="lg:text-sm lg:leading-6 relative">
            <div className="sticky top-0 -ml-0.5 pointer-events-none">
                {!mobile && <div className="h-10 bg-white dark:bg-slate-900" />}
                <div className="bg-white dark:bg-slate-900 relative pointer-events-auto">
                    {/* <SearchButton className="hidden w-full lg:flex items-center text-sm leading-6 text-slate-400 rounded-md ring-1 ring-slate-900/10 shadow-sm py-1.5 pl-2 pr-3 hover:ring-slate-300 dark:bg-slate-800 dark:highlight-white/5 dark:hover:bg-slate-700">
                    {({ actionKey }) => (
                        <>
                        <svg
                            width="24"
                            height="24"
                            fill="none"
                            aria-hidden="true"
                            className="mr-3 flex-none"
                        >
                            <path
                            d="m19 19-3.5-3.5"
                            stroke="currentColor"
                            strokeWidth="2"
                            strokeLinecap="round"
                            strokeLinejoin="round"
                            />
                            <circle
                            cx="11"
                            cy="11"
                            r="6"
                            stroke="currentColor"
                            strokeWidth="2"
                            strokeLinecap="round"
                            strokeLinejoin="round"
                            />
                        </svg>
                        Quick search...
                        {actionKey && (
                            <span className="ml-auto pl-3 flex-none text-xs font-semibold">
                            {actionKey[0]}K
                            </span>
                        )}
                        </>
                    )}
                    </SearchButton> */}
                </div>
                {!mobile && <div className="h-8 bg-gradient-to-b from-white dark:from-slate-900" />}
            </div>
            <ul>
                {nav.map((section) => (
                    <li key={section.title} className="mt-12 lg:mt-8">
                        <h5 className="mb-8 lg:mb-3 font-semibold text-slate-900 dark:text-slate-200">
                            {section.title}
                        </h5>
                        <ul className={clsx(
                            'space-y-6 lg:space-y-2 border-l border-slate-100',
                            mobile ? 'dark:border-slate-700' : 'dark:border-slate-800'
                        )}>
                            {section.links.map((item) => (
                                <li key={item.title}>
                                    <NavItem 
                                        href={item.href}
                                        isActive={item.href === router.asPath.split('#')[0]}
                                        label={item.title}
                                    />
                                </li>
                            ))}
                        </ul>
                    </li>
                ))}
            </ul>
        </nav>
    );
};