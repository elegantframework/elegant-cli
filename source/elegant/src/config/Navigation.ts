import { NavigationSection } from '@/types/Navigation';

/**
 * The configuration for the documentation sidebar navigation.
 */
export const documentationNav: NavigationSection[] = [
  {
    title: "Welcome",
    links: [
      {
        title: "Welcome to Elegant",
        href: "/docs/welcome"
      },
    ]
  },
];