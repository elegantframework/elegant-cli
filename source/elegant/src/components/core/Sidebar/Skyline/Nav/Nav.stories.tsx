import type { Meta, StoryObj } from '@storybook/react';
import Nav from './Nav';
import { documentationNav } from '@/config/Navigation';

const meta: Meta<typeof Nav> = {
  title: 'Core/Sidebar/Skyline/Nav',
  component: Nav
};

export default meta;

type Story = StoryObj<typeof Nav>;

export const Default: Story = {
  args: {
    mobile: false,
    nav: documentationNav
  }
};