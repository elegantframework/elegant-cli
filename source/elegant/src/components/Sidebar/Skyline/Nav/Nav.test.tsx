import React from "react";
import renderer from 'react-test-renderer';
import Nav from "./Nav";
import { documentationNav } from "@/config/Navigation";

describe("Navigation for the sidebar", () => {

    const useRouter = jest.spyOn(require("next/router"), "useRouter");

    useRouter.mockImplementation(() => ({
        route: '/',
        pathname: '',
        query: '',
        asPath: '',
        push: jest.fn(),
        events: {
        on: jest.fn(),
        off: jest.fn()
        },
        beforePopState: jest.fn(() => null),
        prefetch: jest.fn(() => null)
    }));

    it('renders the navigation section as expected', () => {
        const nav = renderer
        .create(
            <Nav nav={documentationNav}/>
        )
        .toJSON();
        expect(nav).toMatchSnapshot();
    });
});