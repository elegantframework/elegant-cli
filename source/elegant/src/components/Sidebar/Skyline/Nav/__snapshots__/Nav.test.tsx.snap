// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navigation for the sidebar renders the navigation section as expected 1`] = `
<nav
  className="lg:text-sm lg:leading-6 relative"
>
  <div
    className="sticky top-0 -ml-0.5 pointer-events-none"
  >
    <div
      className="h-10 bg-white dark:bg-slate-900"
    />
    <div
      className="bg-white dark:bg-slate-900 relative pointer-events-auto"
    />
    <div
      className="h-8 bg-gradient-to-b from-white dark:from-slate-900"
    />
  </div>
  <ul>
    <li
      className="mt-12 lg:mt-8"
    >
      <h5
        className="mb-8 lg:mb-3 font-semibold text-slate-900 dark:text-slate-200"
      >
        Welcome
      </h5>
      <ul
        className="space-y-6 lg:space-y-2 border-l border-slate-100 dark:border-slate-800"
      >
        <li>
          <a
            className="block border-l pl-4 -ml-px border-transparent hover:border-slate-400 dark:hover:border-slate-500"
            href="/docs/welcome"
            onClick={[Function]}
            onMouseEnter={[Function]}
            onTouchStart={[Function]}
          >
            Welcome to Elegant
          </a>
        </li>
      </ul>
    </li>
  </ul>
</nav>
`;
